internal/pkg/checkin/bulk.go

```diff
@@ -179,35 +264,68 @@ func parseOpts(opts ...Opt) optionsT {
 	return outOpts
 }
 
-// Generate and cache timestamp on seconds change.
-// Avoid thousands of formats of an identical string.
-func (bc *Bulk) timestamp() string {
-	// WARNING: Expects mutex locked.
-	now := time.Now()
-	if now.Unix() != bc.unix {
-		bc.unix = now.Unix()
-		bc.ts = now.UTC().Format(time.RFC3339)
-	}
-
-	return bc.ts
+// Add adds the agent to the connected list.
+func (bc *Bulk) Add(id string) {
+	bc.mut.Lock()
+	defer bc.mut.Unlock()
+	bc.connected[id] = struct{}{}
 }
 
-// CheckIn will add the agent (identified by id) to the pending set.
-// The pending agents are sent to elasticsearch as a bulk update at each flush interval.
-// NOTE: If Checkin is called after Run has returned it will just add the entry to the pending map and not do any operations, this may occur when the fleet-server is shutting down.
-// WARNING: Bulk will take ownership of fields, so do not use after passing in.
-func (bc *Bulk) CheckIn(id string, opts ...Option) error {
+// Remove removes the agent from the connected list.
+func (bc *Bulk) Remove(id string) {
 	bc.mut.Lock()
-	pending := pendingT{
-		ts: bc.timestamp(),
-	}
+	defer bc.mut.Unlock()
+	delete(bc.connected, id)
+}
 
+// CheckIn records that the agent has checked-in.
+//
+// This does not call `Add`, the caller should also call that to record that the agent is connected so that bulk
+// can keep the documents `updated_at` fields updated.
+// WARNING: CheckIn will take ownership of fields, so do not use after passing in.
+func (bc *Bulk) CheckIn(ctx context.Context, id string, opts ...Option) error {
+	now := time.Now().UTC().Format(time.RFC3339)
+	checkin := checkinT{
+		ts: now,
+	}
 	for _, opt := range opts {
-		opt(&pending)
+		opt(&checkin)
+	}
+
+	// update the agent document
+	body, err := checkin.toBody()
+	if err != nil {
+		return fmt.Errorf("could not marshall update body: %w", err)
+	}
+	bulkOpts := []bulk.Opt{bulk.WithRetryOnConflict(3)}
+	if checkin.seqNo.IsSet() {
+		bulkOpts = append(bulkOpts, bulk.WithRefresh())
+	}
+	err = bc.bulker.Update(ctx, dl.FleetAgents, id, body, bulkOpts...)
+	if err != nil {
+		return fmt.Errorf("failed to update document: %w", err)
+	}
+
+	// deleteAudit performs a second request (being that this should not happen very often
+	// it's safer to only use this script when it is needed).
+	if checkin.deleteAudit {
```

Comment:
One thing I tried to avoid with the initial implementation was a second ES request to run the script; network calls are slow and reducing calls made/load generally makes fleet more stable.